overlord/devicestate/handlers_systems.go

```diff
@@ -0,0 +1,375 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+func taskRecoverySystemSetup(t *state.Task) (*recoverySystemSetup, error) {
+	var setup recoverySystemSetup
+
+	err := t.Get("recovery-system-setup", &setup)
+	if err == nil {
+		return &setup, nil
+	}
+	if err != state.ErrNoState {
+		return nil, err
+	}
+	// find the task which holds the data
+	var id string
+	if err := t.Get("recovery-system-setup-task", &id); err != nil {
+		return nil, err
+	}
+	ts := t.State().Task(id)
+	if ts == nil {
+		return nil, fmt.Errorf("internal error: cannot find referenced task %v", id)
+	}
+	if err := ts.Get("recovery-system-setup", &setup); err != nil {
+		return nil, err
+	}
+	return &setup, nil
+}
+
+func setTaskRecoverySystemSetup(t *state.Task, setup *recoverySystemSetup) error {
+	if t.Has("recovery-system-setup") {
+		t.Set("recovery-system-setup", setup)
+		return nil
+	}
+	var id string
```

Comment:
this part of the code isn't used or tested ATM, do we need it?

Reply:
Dropped